File: test/dynamics_frf_tests.f90
module dynamics_frf_tests
    use dynamics
    use iso_fortran_env
    use fortran_test_helper
    implicit none

contains
! ------------------------------------------------------------------------------
! 2nd Order Test Problem
! x" + 2 * z * wn * x' + wn**2 * x = f(t)
pure function example_2nd_order_forcing(t) result(rst)
    ! Arguments
    real(real64), intent(in) :: t
    real(real64) :: rst

    ! Process
    rst = chirp(t, 1.0d2, 5.0d0, 1.0d0, 1.0d2)
end function

pure subroutine example_2nd_order(t, x, dxdt)
    ! Arguments
    real(real64), intent(in) :: t
    real(real64), intent(in), dimension(:) :: x
    real(real64), intent(out), dimension(:) :: dxdt

    ! Model Parameters
    real(real64), parameter :: pi = 2.0d0 * acos(0.0d0)
    real(real64), parameter :: z = 1.0d-1
    real(real64), parameter :: wn = 2.0d0 * pi * 5.0d1

    ! Local Variables
    real(real64) :: f

    ! Process
    f = example_2nd_order_forcing(t)
    dxdt(1) = x(2)
    dxdt(2) = f - (2.0d0 * z * wn * x(2) + wn**2 * x(1))
end subroutine

pure subroutine example_2nd_order_sweep(freq, x, y, dydx)
    ! Arguments
    real(real64), intent(in) :: freq
    real(real64), intent(in) :: x
    real(real64), intent(in), dimension(:) :: y
    real(real64), intent(out), dimension(:) :: dydx

    ! Model Parameters
    real(real64), parameter :: pi = 2.0d0 * acos(0.0d0)
    real(real64), parameter :: z = 1.0d-1
    real(real64), parameter :: wn = 2.0d0 * pi * 5.0d1

    ! Local Variables
    real(real64) :: f

    ! Process
    f = sin(2.0d0 * pi * freq * x)
    dydx(1) = y(2)
    dydx(2) = f - (2.0d0 * z * wn * y(2) + wn**2 * y(1))
end subroutine
! ------------------------------------------------------------------------------
function test_frf_sweep() result(rst)
    ! Arguments
    logical :: rst

    ! Parameters
    real(real64), parameter :: fmax = 6.0d1
    real(real64), parameter :: fmin = 4.0d1
    real(real64), parameter :: tol = 0.05d0
    integer(int32), parameter :: npts = 10
    real(real64), parameter :: pi = 2.0d0 * acos(0.0d0)
    real(real64), parameter :: z = 1.0d-1
    real(real64), parameter :: wn = 2.0d0 * pi * 5.0d1
    complex(real64), parameter :: j = (0.0d0, 1.0d0)

    ! Local Variables
    procedure(harmonic_ode), pointer :: fcn
    integer(int32) :: i
    real(real64) :: df, freq(npts)
    real(real64), allocatable, dimension(:) :: mag1, mag2, magans1, magans2, &
        ratio1, ratio2, ref
    type(frf) :: sol
    complex(real64), allocatable, dimension(:) :: tf1, tf2, s

    ! Initialization
    rst = .true.
    fcn => example_2nd_order_sweep
    df = (fmax - fmin) / (npts - 1.0d0)
    freq = (/ (df * i + fmin, i = 0, npts - 1) /)

    ! Compute the FRF's
    sol = frequency_sweep(fcn, freq, [0.0d0, 0.0d0])
    mag1 = abs(sol%responses(:,1))
    mag2 = abs(sol%responses(:,2))

    ! Compute the solution
    s = j * (2.0d0 * pi * freq)
    tf1 = 1.0d0 / (s**2 + 2.0d0 * z * wn * s + wn**2)
    tf2 = s * tf1
    magans1 = abs(tf1)
    magans2 = abs(tf2)

    ! Tests
    ratio1 = mag1 / magans1
    allocate(ref(size(ratio1)), source = 1.0d0)
    if (.not.assert(ratio1, ref, tol)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_frf_sweep -1"
    end if

    ratio2 = mag2 / magans2
    if (.not.assert(ratio2, ref, tol)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_frf_sweep -2"
    end if
end function

! ------------------------------------------------------------------------------
subroutine modal_frf_forcing_term(freq, f)
    real(real64), intent(in) :: freq
    complex(real64), intent(out), dimension(:) :: f

    complex(real64), parameter :: zero = (0.0d0, 0.0d0)
    complex(real64), parameter :: one = (1.0d0, 0.0d0)

    f = [one, zero, zero]
end subroutine

! Use the example from: https://github.com/jchristopherson/linalg
function test_proportional_damping_frf() result(rst)
    ! Arguments
    logical :: rst

    ! Parameters
    real(real64), parameter :: tol = 1.0d-3
    integer(int32), parameter :: nfreq = 100
    real(real64), parameter :: fmin = 10.0d0
    real(real64), parameter :: fmax = 1.0d2
    real(real64), parameter :: alpha = 0.1d0
    real(real64), parameter :: beta = 2.0d-5
    real(real64), parameter :: ans1(3) = [232.9225d0, 749.6189d0, 923.5669d0]

    ! Define the model parameters
    real(real64), parameter :: m1 = 0.5d0
    real(real64), parameter :: m2 = 2.5d0
    real(real64), parameter :: m3 = 0.75d0
    real(real64), parameter :: k1 = 5.0d6
    real(real64), parameter :: k2 = 10.0d6
    real(real64), parameter :: k3 = 10.0d6
    real(real64), parameter :: k4 = 5.0d6

    ! Local Variables
    integer(int32) :: i
    real(real64) :: df, m(3,3), k(3,3), freq(nfreq)
    real(real64), allocatable, dimension(:) :: modes
    real(real64), allocatable, dimension(:,:) :: modeshapes
    procedure(modal_excite), pointer :: fcn
    type(frf) :: rsp

    ! Initialization
    rst = .true.
    df = (fmax - fmin) / (nfreq - 1.0d0)
    freq = (/ (df * i + fmin, i = 0, nfreq - 1) /)
    fcn => modal_frf_forcing_term

    ! Define the mass matrix
    m = reshape([m1, 0.0d0, 0.0d0, 0.0d0, m2, 0.0d0, 0.0d0, 0.0d0, m3], [3, 3])

    ! Define the stiffness matrix
    k = reshape([k1 + k2, -k2, 0.0d0, -k2, k2 + k3, -k3, 0.0d0, -k3, k3 + k4], &
        [3, 3])

    ! Compute the frequency response functions, along with the modal information
    rsp = frequency_response(m, k, alpha, beta, freq, fcn, modes, modeshapes)

    ! Verify the modal results
    if (.not.assert(ans1, modes, tol)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_proportional_damping_frf -1"
    end if
end function

! ------------------------------------------------------------------------------
! Use the example from: https://github.com/jchristopherson/linalg
function test_modal_response() result(rst)
    ! Arguments
    logical :: rst

    ! Parameters
    real(real64), parameter :: tol = 1.0d-3
    integer(int32), parameter :: nfreq = 100
    real(real64), parameter :: fmin = 10.0d0
    real(real64), parameter :: fmax = 1.0d2
    real(real64), parameter :: alpha = 0.1d0
    real(real64), parameter :: beta = 2.0d-5
    real(real64), parameter :: ans1(3) = [232.9225d0, 749.6189d0, 923.5669d0]
    real(real64), parameter :: vec1(3) = [0.7179d0, 1.0d0, 0.7466d0]
    real(real64), parameter :: vec2(3) = [-0.4192d0, -0.1638d0, 1.0d0]
    real(real64), parameter :: vec3(3) = [1.0d0, -0.1837d0, 0.1791d0]

    ! Define the model parameters
    real(real64), parameter :: m1 = 0.5d0
    real(real64), parameter :: m2 = 2.5d0
    real(real64), parameter :: m3 = 0.75d0
    real(real64), parameter :: k1 = 5.0d6
    real(real64), parameter :: k2 = 10.0d6
    real(real64), parameter :: k3 = 10.0d6
    real(real64), parameter :: k4 = 5.0d6

    ! Local Variables
    real(real64) :: m(3,3), k(3,3)
    real(real64), allocatable, dimension(:) :: modes
    real(real64), allocatable, dimension(:,:) :: modeshapes

    ! Initialization
    rst = .true.

    ! Define the mass matrix
    m = reshape([m1, 0.0d0, 0.0d0, 0.0d0, m2, 0.0d0, 0.0d0, 0.0d0, m3], [3, 3])

    ! Define the stiffness matrix
    k = reshape([k1 + k2, -k2, 0.0d0, -k2, k2 + k3, -k3, 0.0d0, -k3, k3 + k4], &
        [3, 3])

    ! Compute the modal response
    call modal_response(m, k, modes, modeshapes)

    ! Normalize the mode shapes
    call normalize_mode_shapes(modeshapes)

    ! Verify the modal frequencies
    if (.not.assert(ans1, modes, tol)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_modal_response -1"
    end if

    ! Test the mode shapes
    if (.not.assert(vec1, modeshapes(:,1), tol)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_modal_response -2"
    end if
    if (.not.assert(vec2, modeshapes(:,2), tol)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_modal_response -3"
    end if
    if (.not.assert(vec3, modeshapes(:,3), tol)) then
        rst = .false.
        print "(A)", "TEST FAILED: test_modal_response -4"
    end if
end function

! ------------------------------------------------------------------------------
end module

File: test/c/dynamics_frf_tests.c
#include "dynamics.h"
#include "dynamics_frf_tests.h"
#include "dynamics_c_test_helper.h"
#include <math.h>
#include <stdio.h>
#include <float.h>

void c_modal_forcing_term(int n, double freq, double complex *f)
{
    const double complex one = 1.0 + 0.0 * I;
    const double complex zero = 0.0 + 0.0 * I;
    f[0] = one;
    f[1] = zero;
    f[2] = zero;
}

bool c_test_frequency_response()
{
    // Local Variables
    bool rst;
    const int nfreq = 100;
    const double pi = 2.0 * acos(0.0);
    const double tol = 1.0e-3;
    const double fmin = 10.0;
    const double fmax = 1.0e2;
    const double alpha = 0.1;
    const double beta = 2.0e-5;
    const double m1 = 0.5;
    const double m2 = 2.5;
    const double m3 = 0.75;
    const double k1 = 5.0e6;
    const double k2 = 10.0e6;
    const double k3 = 10.0e6;
    const double k4 = 5.0e6;
    int i;
    double df, ans1[3], m[9], k[9], freq[nfreq], omega[nfreq], modes[3], 
        shapes[9];
    double complex rsp[3 * nfreq];
    c_modal_excite fcn;

    // Initialization
    rst = true;
    fcn = c_modal_forcing_term;
    ans1[0] = 2.0 * pi * 232.9225;
    ans1[1] = 2.0 * pi * 749.6189;
    ans1[2] = 2.0 * pi * 923.5669;
    df = (fmax - fmin) / (nfreq - 1.0);
    for (i = 0; i < nfreq; ++i)
    {
        freq[i] = df * i + fmin;
        omega[i] = 2.0 * pi * freq[i];
    }

    // Define the mass matrix
    m[0] = m1;
    m[1] = 0.0;
    m[2] = 0.0;

    m[3] = 0.0;
    m[4] = m2;
    m[5] = 0.0;

    m[6] = 0.0;
    m[7] = 0.0;
    m[8] = m3;

    // Define the stiffness matrix
    k[0] = k1 + k2;
    k[1] = -k2;
    k[2] = 0.0;

    k[3] = -k2;
    k[4] = k2 + k3;
    k[5] = -k3;

    k[6] = 0.0;
    k[7] = -k3;
    k[8] = k3 + k4;

    // Compute the FRF's along with the modal information
    c_frequency_response(3, nfreq, m, 3, k, 3, alpha, beta, omega, fcn, modes, 
        shapes, 3, rsp, nfreq);
    if (!compare_arrays(3, ans1, modes, tol))
    {
        rst = false;
        printf("TEST FAILED: c_test_frequency_response -1\n");
    }

    // End
    return rst;
}



bool c_test_modal_response()
{
    // Local Variables
    bool rst;
    double m[9], k[9], ans1[3], vec1[3], vec2[3], vec3[3], freqs[3], shapes[9];
    const double pi = 2.0 * acos(0.0);
    const double m1 = 0.5;
    const double m2 = 2.5;
    const double m3 = 0.75;
    const double k1 = 5.0e6;
    const double k2 = 10.0e6;
    const double k3 = 10.0e6;
    const double k4 = 5.0e6;
    const double tol = 1.0e-3;

    // Initialization
    rst = true;
    ans1[0] = 2.0 * pi * 232.9225;
    ans1[1] = 2.0 * pi * 749.6189;
    ans1[2] = 2.0 * pi * 923.5669;
    
    vec1[0] = 0.7179;
    vec1[1] = 1.0;
    vec1[2] = 0.7466;

    vec2[0] = -0.4192;
    vec2[1] = -0.1638;
    vec2[2] = 1.0;

    vec3[0] = 1.0;
    vec3[1] = -0.1837;
    vec3[2] = 0.1791;

    // Define the mass matrix
    m[0] = m1;
    m[1] = 0.0;
    m[2] = 0.0;

    m[3] = 0.0;
    m[4] = m2;
    m[5] = 0.0;

    m[6] = 0.0;
    m[7] = 0.0;
    m[8] = m3;

    // Define the stiffness matrix
    k[0] = k1 + k2;
    k[1] = -k2;
    k[2] = 0.0;

    k[3] = -k2;
    k[4] = k2 + k3;
    k[5] = -k3;

    k[6] = 0.0;
    k[7] = -k3;
    k[8] = k3 + k4;

    // Compute the modal response
    c_modal_response(3, m, 3, k, 3, freqs, shapes, 3);
    c_normalize_mode_shapes(3, shapes, 3);

    // Tests
    if (!compare_arrays(3, ans1, freqs, tol))
    {
        rst = false;
        printf("TEST FAILED: c_test_modal_response -1\n");
    }
    if (!compare_arrays(3, vec1, shapes, tol))
    {
        rst = false;
        printf("TEST FAILED: c_test_modal_response -2\n");
    }
    if (!compare_arrays(3, vec2, &shapes[3], tol))
    {
        rst = false;
        printf("TEST FAILED: c_test_modal_response -3\n");
    }
    if (!compare_arrays(3, vec3, &shapes[6], tol))
    {
        rst = false;
        printf("TEST FAILED: c_test_modal_response -4\n");
    }


    // End
    return rst;
}



void c_example_2nd_order_sweep(int n, double freq, double t, const double *x,
    double *dxdt)
{
    // Local Variables
    const double pi = 2.0 * acos(0.0);
    const double z = 1.0e-1;
    const double wn = 2.0 * pi * 50.0;
    double f;

    f = sin(2.0 * pi * freq * t);
    dxdt[0] = x[1];
    dxdt[1] = f - (2.0 * z * wn * x[1] + wn * wn * x[0]);
}

bool c_test_frf_sweep()
{
    // Local Variables
    bool rst;
    const double fmax = 60.0;
    const double fmin = 40.0;
    const double tol = 0.05;
    const int npts = 10;
    const double pi = 2.0 * acos(0.0);
    const double z = 1.0e-1;
    const double wn = 2.0 * pi * 50.0;
    c_frequency_sweep_controls opts;
    int i;
    double df, freq[npts], omega[npts], mag1, mag2, magans1[npts],
        magans2[npts], ref[npts], ratio1[npts], ratio2[npts], init_vals[2];
    double complex rsp[npts * 2], s, tf1, tf2;

    // Initialization
    rst = true;
    c_set_frequency_sweep_defaults(&opts);
    init_vals[0] = 0.0;
    init_vals[1] = 0.0;
    df = (fmax - fmin) / (npts - 1.0);
    for (i = 0; i < npts; ++i)
    {
        freq[i] = df * i + fmin;
        omega[i] = 2.0 * pi * freq[i];
        ref[i] = 1.0;
    }

    // Compute the solution
    for (i = 0; i < npts; ++i)
    {
        s = I * omega[i];
        tf1 = 1.0 / (s * s + 2.0 * z * wn * s + wn * wn);
        tf2 = tf1 * s;
        magans1[i] = cabs(tf1);
        magans2[i] = cabs(tf2);
    }

    // Perform the sweep operation
    c_frf_sweep(2, npts, c_example_2nd_order_sweep, freq, init_vals, 
        DYN_RUNGE_KUTTA_45, rsp, npts, &opts);
    for (i = 0; i < npts; ++i)
    {
        mag1 = cabs(rsp[INDEX(i,0,npts)]);
        mag2 = cabs(rsp[INDEX(i,1,npts)]);
        ratio1[i] = mag1 / magans1[i];
        ratio2[i] = mag2 / magans2[i];
    }

    // Test
    if (!compare_arrays(npts, ref, ratio1, tol))
    {
        rst = false;
        printf("TEST FAILED: c_test_frf_sweep -1\n");
    }
    if (!compare_arrays(npts, ref, ratio2, tol))
    {
        rst = false;
        printf("TEST FAILED: c_test_frf_sweep -2\n");
    }

    // End
    return rst;
}


bool c_test_frf_fit()
{
    // Local Variables
    bool rst;
    const double pi = 2.0 * acos(0.0);
    const double tol = 5.0e-2;
    const int nfreq = 100;
    const int order = 3;
    const double fmin = 2.0 * pi * 10.0;
    const double fmax = 2.0 * pi * 1.0e3;
    const double alpha = 0.1;
    const double beta = 2.0e-5;
    const double m1 = 0.5;
    const double m2 = 2.5;
    const double m3 = 0.75;
    const double k1 = 5.0e6;
    const double k2 = 10.0e6;
    const double k3 = 10.0e6;
    const double k4 = 5.0e6;
    int i;
    double m[9], k[9], ans[nfreq], mdl[3 * order], fitamp[nfreq], modes[3],
        modeshapes[9], df, freq[nfreq], maxp[3 * order], minp[3 * order],
        maxval;
    double complex fit[nfreq], nrmrsp[3 * nfreq], val;
    c_modal_excite fcn;
    c_iteration_controls controls;
    c_regression_statistics stats[3 * order];

    // Initialization
    rst = true;
    fcn = c_modal_forcing_term;
    df = (fmax - fmin) / (nfreq - 1.0);
    c_set_iteration_controls_defaults(&controls);
    for (i = 0; i < nfreq; ++i)
    {
        freq[i] = fmin + df * i;
    }
    for (i = 0; i < 3 * order; ++i)
    {
        maxp[i] = DBL_MAX;
        minp[i] = -DBL_MAX;
    }

    // Define the mass matrix
    m[0] = m1;
    m[1] = 0.0;
    m[2] = 0.0;

    m[3] = 0.0;
    m[4] = m2;
    m[5] = 0.0;

    m[6] = 0.0;
    m[7] = 0.0;
    m[8] = m3;

    // Define the stiffness matrix
    k[0] = k1 + k2;
    k[1] = -k2;
    k[2] = 0.0;

    k[3] = -k2;
    k[4] = k2 + k3;
    k[5] = -k3;

    k[6] = 0.0;
    k[7] = -k3;
    k[8] = k3 + k4;

    // Compute the frequency response functions
    c_frequency_response(3, nfreq, m, 3, k, 3, alpha, beta, freq, fcn,
        modes, modeshapes, 3, nrmrsp, nfreq);

    // Normalize the response
    val = nrmrsp[0];
    for (i = 0; i < nfreq; ++i)
    {
        nrmrsp[i] /= val;
    }

    // Fit the response
    c_fit_frf(nfreq, order, DYN_RECEPTANCE_MODEL, freq, nrmrsp, maxp, minp,
        &controls, mdl, stats);
    c_evaluate_receptance_frf_model(nfreq, order, mdl, freq, fit);

    // Test
    maxval = 0.0;
    for (i = 0; i < nfreq; ++i)
    {
        fitamp[i] = cabs(fit[i]);
        ans[i] = cabs(nrmrsp[i]);
        maxval = maxval < ans[i] ? ans[i] : maxval;
    }
    if (!compare_arrays(nfreq, fitamp, ans, tol * maxval))
    {
        rst = false;
        printf("TEST FAILED: c_test_frf_fit -1\n");
    }

    // End
    return rst;
}
